### PR TITLE
Add CertificateAuthority

### PR DESCRIPTION
**Pull request by DarkaMaul**
#### Summary

Following up on #1182, this PR adds a `CertificateAuthority` class to the trust and uses it to retrieve both Fulcio certificates and Timestamping Authorities.

### CODE REVIEW

File: sigstore/_internal/trust.py [hunk 1]
```diff
@@ -217,6 +220,102 @@ def __str__(self) -> str:
...
+    """
+
+    def __init__(self, inner: _CertificateAuthority):
+        """
+        Construct a new `CertificateAuthority`.
+
+        @api private
+        """
+        self._inner = inner
+        self._leaf: Certificate | None = None
```
Comment (woodruffw) [inline sigstore/_internal/trust.py:235]: Hmm, where is the concept of a "leaf" in a `CertificateAuthority` coming from? The message itself only defines `cert_chain` as being a list of root and intermediate CAs, so there's no "leaf" certificate per se (except in the sense that one of the intermediates is the last one).

I think we should drop this and its associated property:

```suggestion
```

File: sigstore/_internal/trust.py [hunk 2]
```diff
@@ -217,6 +220,102 @@ def __str__(self) -> str:
...
+        Root certificate of the authority.
+        """
+        return self._root
+
+    @property
+    def intermediates(self) -> list[Certificate]:
+        """
+        List of intermediates certificates.
+        """
+        return self._intermediates
```
Comment (woodruffw) [inline sigstore/_internal/trust.py:293]: Related: I think we can remove both of these, and allow access only through the `certificates(...)` method. FWICT they're only used in tests anyways.

**Review by woodruffw — COMMENTED** [review]
Thanks @DarkaMaul! This is a good start, but I think it can be simplified quite a bit -- we don't need to be so stringent about classifying the different members of the `CertificateAuthority` message, since they all end up in a trust "pool" anyways for chain building purposes 🙂

**Comment by DarkaMaul** [comment]
> Thanks @DarkaMaul! This is a good start, but I think it can be simplified quite a bit -- we don't need to be so stringent about classifying the different members of the `CertificateAuthority` message, since they all end up in a trust "pool" anyways for chain building purposes 🙂

I've updated the code to remove the distinction between all the different certificates 👍

**Review by woodruffw — APPROVED** [review]
Nice work!

**Pull request merged**